third_party/amd/lib/TritonAMDGPUTransforms/RescheduleOps.cpp

```diff
@@ -1406,60 +1582,281 @@ SchedDagNode *getOriginalOrder(SchedDagNode *a, SchedDagNode *b) {
+  LocalLoadOps. Here we just want to continue the critical path to specify what
+  is the optimal order of LocalStoreOps, and by consequence the optimal order
+  of LoadOps. Therefore we just want to ensure that LoadOps have the same order.
+
+  TODO(dtanner) Expand this to direct-to-lds.
+  TODO(dtanner) Update this for removing gpu.barriers.
+    For this we will take a LocalLoadOp, determine the dependent LocalStoreOps
+    and continue that propagation.
+******************************************************************************/
+struct LocalStoreCriticalPathPriorityCalculator : public PriorityCalculator {
+
+  LocalStoreCriticalPathPriorityCalculator()
+      : PriorityCalculator(SchedDagNodePriorityType::LocalStoreCriticalPath) {}
+
+  void calcPriorities() {
+    SchedDagNodePriorityDataType priority = 1;
+    for (auto it = dag->nodeList.rbegin(); it != dag->nodeList.rend(); ++it) {
+      SchedDagNode *node = *it;
+      if (isa<triton::gpu::LocalStoreOp>(node->getOp())) {
+        node->setPriority(SchedDagNodePriorityType::LocalStoreCriticalPath,
+                          priority);
+        priority += 1;
+      }
+    }
+
+    // Propagate high priorities up for critical path.
+    for (auto *node : dag->nodeList) {
+      if (isa<triton::gpu::LocalStoreOp>(node->getOp())) {
+        node->propagateHigherPriorityToParents(
+            SchedDagNodePriorityType::LocalStoreCriticalPath);
+      }
+    }
+
+    // Propagate low priorities down for critical path.
+    for (auto it = dag->nodeList.rbegin(); it != dag->nodeList.rend(); ++it) {
+      SchedDagNode *node = *it;
+      if (isa<triton::gpu::LocalStoreOp>(node->getOp())) {
+        node->propagateLowerPriorityToChildren(
+            SchedDagNodePriorityType::LocalStoreCriticalPath);
+      }
+    }
+  }
+};
+
 /*
-  Abstract Base class for a scheduling heuristic recipe.
-  E.g. TopDown, schedule global_loads early and local_stores late.
+  Abstract Base class for scheduling heuristics.
+  E.g. TopDown, schedule LoadOps early and LocalStoreOps late.
 */
 template <SchedDirection Direction> struct SchedHeuristic {
+  SchedHeuristic(StringRef name) : name(name) {}
+  // Scheduler prints the name of heuristic.
+  StringRef getName() const { return name; }
+  // Scheduler calls before beginning a scheduling pass.
+  virtual void begin() {};
+  // Scheduler calls comparison functor to evaluate ready list.
   virtual SchedDagNode *operator()(SchedDagNode *a, SchedDagNode *b) = 0;
-  virtual StringRef name() = 0;
+  // Scheduler calls this printer right before evaluating the ready list,
+  // for debugging when the heuristic has state.
+  virtual void dump(llvm::raw_ostream &out) {};
+  // Scheduler allows heuristic to update state based on scheduled op.
+  virtual void selectedOp(SchedDagNode *) {};
+  // Scheduler notifies scheduling is done, for debugging to print final state.
+  virtual void end() {};
   virtual ~SchedHeuristic() = default;
```

Comment:
could you, please, move the destructor to the top? The constructors/destructors are usually at the top of a class/struct

Reply:
Done.